tests/unit/apps/test_auxiliary.py

```diff
@@ -0,0 +1,154 @@
+#  Copyright 2023 Canonical Limited
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Auxiliary application class."""
+import pytest
+
+from cou.apps.auxiliary import AuxiliaryOpenStackApplication
+from cou.exceptions import ApplicationError, HaltUpgradePlanGeneration
+from cou.utils.openstack import OpenStackRelease
+from tests.unit.apps.utils import assert_plan_description
+
+
+def test_auxiliary_app(status, config):
+    # version 3.8 on rabbitmq can be from ussuri to yoga. In that case it will be set as yoga.
+    target = "victoria"
+    expected_units = {"rabbitmq-server/0": {"os_version": "yoga", "workload_version": "3.8"}}
+    app = AuxiliaryOpenStackApplication(
+        "rabbitmq-server",
+        status["rabbitmq_server"],
+        config["auxiliary_ussuri"],
+        "my_model",
+        "rabbitmq-server",
+    )
+    assert app.channel == "3.8/stable"
+    assert app.os_origin == "distro"
+    assert app.units == expected_units
+    assert app.os_origin_config(OpenStackRelease(target)) == "ussuri"
+
+    plan = app.generate_upgrade_plan(target)
+
+    steps_description = [
+        f"Upgrade software packages of '{app.name}' from the current APT repositories",
+        f"Refresh '{app.name}' to the latest revision of '3.8/stable'",
+        f"Change charm config of '{app.name}' 'source' to 'cloud:focal-victoria'",
+        f"Check if the workload of '{app.name}' has been upgraded",
+    ]
+
+    assert_plan_description(plan, steps_description)
```

Comment:
Can we move plan generation testing to a separate function or group all the assert together? The current layout is an anti-pattern (should be "arrange-act-assert "instead of "arrange-act-assert-act-assert")?

Reply:
I've separated into two unit tests